src/yz_wm_search.erl

```diff
@@ -107,24 +105,28 @@ forbidden(RD, Ctx) ->
                                {?YZ_SECURITY_INDEX, wrq:path_info(index, RD)})
     end.
 
-%% Treat POST as GET in order to work with existing Solr clients.
-process_post(Req, S) ->
-    case search(Req, S) of
-        {Val, Req2, S2} when is_binary(Val) ->
-            Req3 = wrq:set_resp_body(Val, Req2),
-            {true, Req3, S2};
-        Other ->
-            %% In this case assume Val is `{halt,Code}' or
-            %% `{error,Term}'
-            Other
+%% @doc Handle POST to work with existing Solr clients and long queries
+%%      when the correct Content-Type is supplied.
+process_post(RD, Ctx) ->
+    CType = wrq:get_req_header(?HEAD_CTYPE, RD),
+    case CType of
+        "application/x-www-form-urlencoded" ->
+            post_search(RD, Ctx, CType);
+        _ ->
+            {{halt, 415}, RD, Ctx}
     end.
 
 search(Req, S) ->
+    Params = wrq:req_qs(Req),
+    search(Req, S, Params).
+
+search(Req, S, []) ->
+    search(Req, S);
```

Comment:
I guess this recursion to arity 2 function is for empty body in POST
search request to fallback to search by qeury string.  When GET search
request without query string (in HTTP sense) or POST one with empty
body and empty query string, then infinite recursion between arity 2
and 3.  Such requests are useless, but I don't know users won't do
such things.

Sample requests:
```
## GET
% curl -s 'http://127.0.0.1:8098/search/query/people'

## POST
% curl -v -X POST -H 'Content-Type: application/x-www-form-urlencoded' \
     -s 'http://127.0.0.1:8098/search/query/people'
```